src/include/sof/audio/sink_api.h

```diff
@@ -0,0 +1,263 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2023 Intel Corporation. All rights reserved.
+ *
+ */
+
+#ifndef __SOF_SINK_API_H__
+#define __SOF_SINK_API_H__
+
+#include <sof/common.h>
+#include <sof/audio/format.h>
+#include <errno.h>
+
+/**
+ * this is a definition of API to sink of audio data
+ *
+ * THE SINK is any component that can store data somehow and provide a buffer to be filled
+ * with data at request. The sink API does not define how the data will be processed/used
+ *
+ * Examples of data sinks:
+ * - /dev/null
+ *     all the data stored in sink buffer are just simply discarded
+ * - I2S sender
+ *     Data stored in sink buffer will be sent to the external world
+ * - a memory ring buffer
+ *     data stored in the buffer will be sent to another module (usually using source API, but it
+ *     does not matter in fact).
+ *
+ *  The main advantage of using source API instead of just taking pointers to the buffers is that
+ *  the buffer may be prepared at the moment the data producer is requesting it. i.e.
+ *   - cache may be written back/invalidated if necessary
+ *   - data may be moved to make linear space
+ *   - part of the buffer may be locked to prevent reading
+ *   etc.etc. it depends on implementation of the data sink
+ *
+ * NOTE: the module should get a complete portion of space it needs for processing, fill it
+ *       than release. The reason is - the depending on the implementation, the calls may be
+ *       expensive - may involve some data moving in memory, cache writebacks, etc.
+ */
+
+/* sink structure obsfucator */
+#define audio_sink_hnd void*
+
+struct sink_ops {
+	/**
+	 * Retrieves size of free space available in sink (in bytes)
+	 * return number of free bytes in buffer available to immediate filling
+	 */
+	uint32_t (*get_free_size)(audio_sink_hnd sink_handler);
+
+	/**
+	 * Get a circular buffer to operate on (to write).
+	 *
+	 * Retrieves a fragment of circular data to be used by the caller
+	 * After calling get_buffer, the space for data is guaranteed to be available
+	 * for exclusive use on the caller core through cached pointer
+	 * The caller MUST take care of data circularity based on provided pointers
+	 *
+	 * @param [in] req_size requested size of space
+	 * @param [out] data_ptr a pointer to the space will be provided there
+	 * @param [out] buffer_start pointer to circular buffer start
+	 * @param [out] buffer_end pointer to circular buffer end
+	 * @param [out] buffer_size size of circular buffer
+	 *
+	 * @retval -ENODATA if req_size is bigger than free space
+	 *
+	 */
+	int (*get_buffer)(audio_sink_hnd src_handler, uint32_t req_size, void **data_ptr,
+			  void **buffer_start, void **buffer_end, uint32_t *buffer_size);
```

Comment:
I don't think these should be exposed. This should be implementation internal. You should only export things like read data, write data, get available data amount, get free space etc. No pointers at all! I think our existing `audio_stream_produce()` / `audio_stream_consume()` are a good direction to follow.

Reply:
That my comment was wrong. We do need pointers because an alternative would be copying data every time for module processing and that would be too expensive. Currently those pointers are obtained by `audio_stream_get_rptr()` and `audio_stream_get_wptr()`. But I'd still remove the last parameter